torch/csrc/jit/codegen/cuda/lower_index.cpp

```diff
@@ -840,6 +840,60 @@ void IndexLowering::generate(const std::vector<Expr*>& exprs) {
   }
 }
 
+kir::Allocate* IndexLowering::allocateUniqueBuffer(
+    Val* buffer_size,
+    DataType dtype,
+    bool zero_init,
+    TensorView* out_tv,
+    std::unordered_map<TensorView*, kir::Allocate*>& alloc_map) {
+  // Return an existing allocation if exists
+  auto it = alloc_map.find(out_tv);
+  if (it != alloc_map.end()) {
+    return it->second;
+  }
+
+  // No existing allocation found. Create a new one
+  auto new_buffer =
+      ir_utils::allocGlobalBufferForGridComm(buffer_size, dtype, zero_init);
+
+  // Keep track of the allocation
+  alloc_map.emplace(out_tv, new_buffer);
+
+  pushBack(new_buffer);
```

Comment:
Is this pushing the allocation within some for loops? Since it's a global allocation I guess it doesn't matter where we put it as the scope is global to the kernel, but might be nice to insert it at a "syntactically correct" location.

Reply:
Yeah, it doesn't matter, but it's not strictly right. I modified to place it at the top-level scope as there's already a function doing that (`insertAtTopLevel`).